Server tests

hasan7n opened the pull request:
This PR adds tests for all server endpoints, focusing on permissions / disallowed actions.


TODO: update [this](https://docs.google.com/spreadsheets/d/1Y-fugi1Y6swtUfC91W1zg3HAH2Yi4Xr7y_nGFlYH7HU/edit#gid=273022396). EDIT: done

closes #509

hasan7n committed:
write all tests

hasan7n committed:
change: remove unused endpoints
These endpoints are not used, and don't seem to help
in our workflow. No need to keep maintaining them

hasan7n committed:
change: fix PUT user permission
PUT now is only permitted to the admin users
We need to define what can be edited and how to sync with
Auth0 before exposing this to the end users

hasan7n committed:
change: fix PUT result behaviour
Users were able to edit results dict and associated entities

hasan7n committed:
change: remove URLs from cube uniqueness criteria

hasan7n committed:
change: apply mlcube rules on both PUT and POST
users have to include hashes for URLs on mlcube creation
but they were able to set it blank later

hasan7n committed:
improve readability

hasan7n committed:
change: modify how associations list works
make dataset association private, return necessary fields
to help the client know the latest approval status

hasan7n committed:
change: fix default approval status

hasan7n committed:
change: fix approved_at update
this field was updated for any request containing approval_status

hasan7n committed:
add notes and TODOs

hasan7n committed:
update workflow file

hasan7n committed:
modify client behaviour for model association list

hasan7n committed:
add some documentation for writing tests

hasan7n committed:
update workflow file

hasan7n committed:
update server readme

hasan7n committed:
Merge remote-tracking branch 'upstream/main' into server-tests

hasan7n committed:
empty

aristizabal95 commented on cli/medperf/comms/rest.py:
```diff
@@ -183,8 +183,10 @@ def get_benchmark_models(self, benchmark_uid: int) -> List[int]:
         """
-        models = self.__get_list(f"{self.server_url}/benchmarks/{benchmark_uid}/models")
-        model_uids = [model["id"] for model in models]
+        # TODO: filter to find approved ones only if we decide to have
```
Should we handle filtering within the comms layer? Or within the entity layer? I think we've done filtering on the entity layer so far, because we also need to apply filtering to local entities

hasan7n replied:
It's tricky for associations since we don't have an association `Entity` on the client yet.

Filtering associations requires two steps:
1. first, finding the latest association of each mlcube. Ideally this should live in an association `Entity`, but for now I think this should live in the comms layer, since the client doesn't actually care about the history of associations. Also, this logic is required for all places in the client code where this comms function is called.
2. Then, ordinary filtering by approval status. Currently this filtering is required in two places:
  a. `entity.list` command: which already applies filtering in the commands layer.
  b. `benchmark.get_models_uids`: which by definition should return approved associations model IDs. We can apply filtering inside this function.

I think this will become cleaner if we decide to create an association `Entity` in our code base.

hasan7n committed:
change: fix mlcube associations list permissions

hasan7n committed:
change: modify result PUT and DELETE
- added is_valid and user_metadata fields
- PUT and DELETE now are only for admin

hasan7n committed:
change: fix permissions of approving the benchmark

hasan7n committed:
change: prevent updating a benchmark to PENDING

hasan7n committed:
Revert "change: remove unused endpoints"

This reverts commit 152862aefc912151a8aeecadda3628d2d1f5865f.
and deactivates the endpoints

hasan7n committed:
change: restrict marking benchmark as operation
doing this requires its three mlcubes to be operation

hasan7n committed:
change: demo dataset url is required

hasan7n committed:
change: fix delete permissions for associations

hasan7n committed:
change: server check for dataset assoc prep mlcube

hasan7n committed:
change: client side new association logic

hasan7n committed:
change: make client side demo url required

hasan7n committed:
cleanup TODOs and related minor changes

hasan7n committed:
fix linter issues

hasan7n committed:
use a better datetime parser

hasan7n committed:
change integration tests order of model priority

hasan7n committed:
stop using benchmark.models for permission reasons

hasan7n committed:
add Result's new fields to the client

hasan7n committed:
refactor to fix linter issue

aristizabal95 commented on server/benchmark/models.py:
```diff
@@ -20,7 +20,7 @@ class Benchmark(models.Model):
     docs_url = models.CharField(max_length=100, blank=True)
     owner = models.ForeignKey(User, on_delete=models.PROTECT)
-    demo_dataset_tarball_url = models.CharField(max_length=256, blank=True)
+    demo_dataset_tarball_url = models.CharField(max_length=256)
```
I remember we talked about this before, but can you remind me how we're getting past situations where demo datasets can't be shared directly with a link?

hasan7n replied:
currently we support using synapse storage (e..g, Verena used a private demo dataset)
we should eventually support things like dropbox or google drive

aristizabal95 reviewed (approved):
Hey Hasan! Just left a few comments/questions for the future. I skimmed over the server tests and implementation. Everything looks great and the tests look very well organized. Amazing amazing work as always Hasan!

hasan7n commented on cli/medperf/tests/entities/test_benchmark.py:
```diff
@@ -20,31 +25,36 @@ def setup(request, mocker, comms, fs):
-        assert benchmark.reference_model_mlcube in benchmark.models
-
-    def test_benchmark_includes_additional_models_in_models(self, setup):
+    def test_benchmark_get_models_works_as_expected(self, setup, expected_models):
```
that's a very valid point
